### ISSUE TITLE
batchFunc instead of gfor?

### ISSUE THREAD

**Issue by unbornchikken** [arrayfire/arrayfire#768]
Hi,

You've mentioned that in another issue that `gfor` is not the right way of doing `gfor` :), gotta use `batchFunc` instead. So, what is `batchFunc`? It's not int the documentation. Can I see an example? What's the status of `gfor`, is it considered as deprecated?

Thanks.

**Comment by pavanky** [comment]
@unbornchikken gfor is not deprecated, but it is also hard to port it to new languages (especially ones without macro support). The essence of gfor or batchFunc is to enable support batched operations. There are many cases that can be termed as batched operations.

**Case 1: Same operation on all elements**

For example if you want to do an operation on each element of array `A` you could do something like this.

```
gfor (seq ii, A.elements())  {
B(ii) = sin(A(ii));
}
```

However since `sin` does operations on batches natively, you can just do the following.
```
B = sin(A)
```

Almost all functions support such batched operations include `transpose`, `sort`, `sum` etc. In cases like these using `gfor` is not needed.

**Case 2: Inputs with different batch sizes**

You have an array  that is batched (Iet us say A which is 3D) and another that is not batched (let us say B which is 2D). You may want to add A and B in such a way that each slice in A is added to B to generate the result which is 3D. In such a case you can do one of the following.

Doing the following will cause an error saying inputs are not of the same size.

```
C = A + B;
```

However the following will work:

```
gfor(seq ii, A.dims(2)) {
C(span, span, ii) = A(span, span, ii) + B;
}
```

This function can also be implemented using batch func in the following manner.

```
array add(const array &lhs, const array &rhs)
{
     return lhs + rhs;
}
```

```
C = batchFunc(A, B, add);
```

batchFunc temporarily disables the dimension check and performs a one to all operation and provides the same output as GFOR.

However some functions like `convolve2` already support this natively. This means you do not even need to use `batchFunc` in those cases!

i.e.

```
gfor (seq ii, A.dims(2)) {
 C(span, span, ii) = convolve2(A(span, span, ii), B);
}
```

is the same as:

```
C = convolve2(A, B);
```

In short, gfor exists because it is an "easy to understand" tool for people coming from a non parallel programming background. In arrayfire do not *need* to use it. There is native support for batched operations which is slightly faster than `gfor` in many cases.

**Comment by unbornchikken** [comment]
excellent clarification, thanks. maybe it is a good idea to put this into the wiki, isn't it? btw in javascript i have macros by using eval.

**Comment by unbornchikken** [comment]
oh, and i think that this auto batching functionality should be documented on functions those support it, shouldn't it?

**Comment by unbornchikken** [comment]
Please ignore my last comment, it was from my cellphone, I didn't check docs of convolve2. I can see now that convolve2 documents this behavior pretty well.
